R/utils.R

```diff
@@ -104,3 +104,67 @@ select_apply <- function(fun = "apply") {
 mutate_all_types <- function(dat) {
   return(dat %>% dplyr::mutate_all(utils::type.convert, as.is = TRUE))
 }
+
+#' Register a cluster for parallel processing
+#'
+#' @inheritParams run_ml
+#' @param setup_timeout Setup timeout in seconds. See \link[parallel]{makePSOCKcluster} for details.
+#'
+#' @return PSOCK cluster, or NULL if only 1 core provided or required packages aren't available
+#' @export
+#' @author Kelly Sovacool, \email{sovacool@@umich.edu}
+#'
+#' @examples
+#' para_cluster <- setup_parallel(2)
+#' # insert code that uses foreach here
+#' stop_parallel(para_cluster)
+setup_parallel <- function(ncores, setup_timeout = 0.5) {
+  pcluster <- NULL
+  if (!is.numeric(ncores) & !is.na(ncores)) {
+    warning(paste(
+      "`ncores` must be `NA` or a number, but you provided", ncores,
+      "\nProceeding with only one process."
+    ))
+  } else if (!is.na(ncores) & ncores > 1) {
+    if (!all(check_package_installed(c("parallel", "doParallel", "foreach")))) {
+      warning(paste(
+        "The packages `parallel`, `doParallel`, and `foreach` are required for using multiple cores.\n",
+        "You specified", ncores, "cores, but one or more of these packages are not installed.\n",
+        "Proceeding with only one process."
+      ))
+    } else {
+      cores_avail <- parallel::detectCores()
+      if (ncores > cores_avail) {
+        warning(paste(
+          "You specified", ncores, "cores, but only", cores_avail, "cores are available.",
+          "\nProceeding with only one process."
```

Comment:
For this should we proceed with the number of detected cores instead of one process?

Reply:
Initially I thought about going with `cores_avail - 1` or `cores_avail / 2`. However, `R CMD CHECK` only supports up to 2 cores at a time, so it fails on GitHub actions if tests try to use more than 2. We also can't assume that the user won't have other jobs running on the other half of the cores requested. So I thought it'd be better to just force the user to pick a good number if they want parallelization.